Code/GraphMol/FMCS/Wrap/rdFMCS.cpp

```diff
@@ -11,41 +11,383 @@
+};
+
+struct PyProgressCallbackUserData {
+  const MCSProgressData *mcsProgressData;
+  python::object pyMCSProgress;
+  python::object pyAtomComp;
+  python::object pyBondComp;
+};
+
+struct PyMCSProgress {
+  virtual bool callback(const MCSProgressData&,
+                        const MCSParameters&) {
+    PyErr_SetString(PyExc_AttributeError,
+      "The " CALLBACK_FUNC_NAME "() method "
+      "must be overridden in the rdFMCS.MCSProgress subclass");
+    python::throw_error_already_set();
+    return false;
+  }
+};
+
+class PyMCSProgressData {
+public:
+  PyMCSProgressData() :
+    pd(new MCSProgressData()),
+    pcud(new PyProgressCallbackUserData()) {
+      pcud->mcsProgressData = pd.get();
+    }
+  PyMCSProgressData(const MCSProgressData &other) :
+    PyMCSProgressData() {
+      *pd = other;
+    }
+  unsigned int getNumAtoms() const {
+    return pd->NumAtoms;
+  }
+  unsigned int getNumBonds() const {
+    return pd->NumBonds;
+  }
+  unsigned int getSeedProcessed() const {
+    return pd->SeedProcessed;
+  }
+private:
+  std::unique_ptr<MCSProgressData> pd;
+  std::unique_ptr<PyProgressCallbackUserData> pcud;
+};
+
+class PyMCSParameters {
+public:
+  PyMCSParameters() :
+    p(new MCSParameters()),
+    cfud(new PyCompareFunctionUserData()),
+    pcud(new PyProgressCallbackUserData()) {
+      cfud->mcsParameters = p.get();
+      pcud->mcsProgressData = nullptr;
+    }
+  PyMCSParameters(const MCSParameters &other,
+                  const PyProgressCallbackUserData &pcudOther) :
+    PyMCSParameters() {
+      *p = other;
+      pcud->pyMCSProgress = pcudOther.pyMCSProgress;
+      cfud->pyAtomComp = pcudOther.pyAtomComp;
+      cfud->pyBondComp = pcudOther.pyBondComp;
+    }
+  const MCSParameters *get() const {
+    return p.get();
+  }
+  bool getMaximizeBonds() const {
+    return p->MaximizeBonds;
+  }
+  void setMaximizeBonds(bool value) {
+    p->MaximizeBonds = value;
+  }
+  double getThreshold() const {
+    return p->Threshold;
+  }
+  void setThreshold(double value) {
+    p->Threshold = value;
+  }
+  unsigned int getTimeout() const {
+    return p->Timeout;
+  }
+  void setTimeout(unsigned int value) {
+    p->Timeout = value;
+  }
+  bool getVerbose() const {
+    return p->Verbose;
+  }
+  void setVerbose(bool value) {
+    p->Verbose = value;
+  }
+  const MCSAtomCompareParameters& getAtomCompareParameters() const {
+    return p->AtomCompareParameters;
+  }
+  void setAtomCompareParameters(const MCSAtomCompareParameters &value) {
+    p->AtomCompareParameters = value;
+  }
+  const MCSBondCompareParameters& getBondCompareParameters() const {
+    return p->BondCompareParameters;
+  }
+  void setBondCompareParameters(const MCSBondCompareParameters &value) {
+    p->BondCompareParameters = value;
+  }
+  std::string getInitialSeed() const {
+    return p->InitialSeed;
+  }
+  void setInitialSeed(const std::string &value) {
+    p->InitialSeed = value;
+  }
+  void setMCSAtomTyper(PyObject *atomComp) {
```

Comment:
It's maybe not technically necessary here, but I always feel safer if a function taking a pointer that is going to be dereferenced has a `PRECONDITION()` on that pointer.